### DISCUSSION TITLE
How can I set up <j> and <k> to navigate up and down, and make <i> to focus on the search?

### DISCUSSION THREAD

**Discussion by wpsnapper** [ibhagwan/fzf-lua#1653] [Q&A] [answered]
I've been enjoying fzf.lua and it's working really good for me. However, I can't replicate what I had with telescope. In telescope I used <jk> buttons to go up and down and when I need to focus on the search again I'd press <i> and I can search again. I press <ESC> and move back and forth again if I want.

I added the following code, but it doesn't let me focus back on the search input or type a name containing those two letters ('jk').

```
require('fzf-lua').setup({
  keymap = {
    fzf = {
      true,
      ["h"] = "abort",
      ["j"] = "down",
      ["k"] = "up",
      ["l"] = "accept",
    },
  }
})
```

Is this possible to do with fzf.lua in anther way?

**Comment by ibhagwan** [comment]
The below sets this up to use `i` for the fake "insert" mode and `esc` to go back to the fake "normal" mode:
> For more info, see `man fzf` for understaning `bind|unbind`
```lua
require('fzf-lua').setup({
  keymap = {
    fzf = {
      true,
      ["h"] = "abort",
      ["j"] = "down",
      ["k"] = "up",
      ["l"] = "accept",
      ["i"] = "unbind(h)+unbind(j)+unbind(k)+unbind(l)+unbind(i)",
      ["esc"] = "rebind(h)+rebind(j)+rebind(k)+rebind(l)+rebind(i)",
    },
  }
})
```

**Reply by wpsnapper** [reply]
Thanks for getting back to me! For some reason, adding those keystrokes made my fzf feel sluggish. I think I'll just revert it for now and stick to using <shift><j> and <shift><k> for down and up.

I think people would really appreciate it if you added this feature natively—it makes navigation so much smoother without having to move your fingers off the keys. That said, I’ll keep using fzf.lua even without it because it’s just so snappy with large codebases. People like you make the world a better place—thanks for all your work!

**Reply by ibhagwan** [reply]
> Thanks for getting back to me! For some reason, adding those keystrokes made my fzf feel sluggish. I think I'll just revert it for now and stick to using and for down and up.

Shouldn’t happen, these are all native fzf keybinds (happening in the term, unrelated to fzf-lua), if you can easily reproduce and demonstrate it would be worthwhile to open an issue upstream.

> I think people would really appreciate it if you added this feature natively—it makes navigation so much smoother without having to move your fingers off the keys. That said, I’ll keep using fzf.lua even without it because it’s just so snappy with large codebases. People like you make the world a better place—thanks for all your work!

I beg to differ, that’s a personal preference, in 2+ years of this plugin existed the “normal” mode request came up only a few times (and not necessarily for hjkl navigation), in addition, this would be better served in the upstream but the maintainer is against that: https://github.com/junegunn/fzf/issues/384#issuecomment-149072848.

If you wish to keep your hands on the home row why don’t you set the same binds with `ctrl-` prefix? I personally use `ctrl-jk` for up|down:
```lua
require('fzf-lua').setup({
  keymap = {
    fzf = {
      true,
      ["ctrl-h"] = "abort",
      ["ctrl-j"] = "down",
      ["ctrl-k"] = "up",
      ["ctrl-l"] = "accept",
    },
  }
})
```

**Comment by ibhagwan** [comment, marked as answer]
FYI, after the latest commit https://github.com/ibhagwan/fzf-lua/commit/12ac915c26ec6337ff96be0b8b2311736cca4c13 you can do the below and keep your `esc` as double purpose to abort from "normal" mode, this also changes the prompt info (e.g. "Files") to be displayed in the title since we lose the prompt:
```lua
require('fzf-lua').setup({
  "default-title", -- Important as we lose the prompt
  keymap = {
    fzf = {
      true,
      ["h"] = "abort",
      ["j"] = "down",
      ["k"] = "up",
      ["l"] = "accept",
      ["i"] = "unbind(h)+unbind(j)+unbind(k)+unbind(l)+unbind(i)+change-prompt(> )",
      ["esc"] = 'transform:case "$FZF_PROMPT" in\n'
          .. '*NORMAL*) echo abort;;\n'
          .. '*) echo "change-prompt(NORMAL - PRESS \'I\' FOR INSERT MODE)'
          .. '+rebind(h)+rebind(j)+rebind(k)+rebind(l)+rebind(i)";;'
          .. '\nesac',
      ["start"] = "change-prompt(NORMAL - PRESS 'I' FOR INSERT MODE)"
    },
  }
})
```

**Comment by wpsnapper** [comment]
Thank you so much. I will try this one.

**Reply by ibhagwan** [reply]
> Thank you so much. I will try this one.

Let me know what you think after you try it out for a while, if this is an acceptable workaround I might add a new profile with it maybe we call it `telescope+normal`?

**Reply by MoulyT** [reply]
Hi! Thanks for taking the time to work on this mode - really appreciate the effort!

I've been testing the configuration but unfortunately ran into an issue with escape sequences. I have both insert mode (jj/jk to <Esc>) and terminal mode (jj/jk to <C-\><C-n>) mappings through better-escape, and I noticed that when these sequences trigger, the cursor escapes fzf's control, just moving through characters without triggering fzf navigation.

I did some testing and found this isn't specific to better-escape - it happens the same way with traditional vim mappings. Either the insert mode or terminal mode escape sequence is enough to cause this behavior.

Hope this helps somehow,
Thanks for considering this and for all your amazing work on fzf-lua!

**Reply by ibhagwan** [reply]
> I've been testing the configuration but unfortunately ran into an issue with escape sequences. I have both insert mode (jj/jk to ) and terminal mode (jj/jk to <C->) mappings through better-escape, and I noticed that when these sequences trigger, the cursor escapes fzf's control, just moving through characters without triggering fzf navigation.

> I did some testing and found this isn't specific to better-escape - it happens the same way with traditional vim mappings. Either the insert mode or terminal mode escape sequence is enough to cause this behavior.

That's because these mappings I provided are mapped in fzf (terminal) while the better-escape mappings are neovim window mappings which then override the fzf mappings as the keys never get to the shell, to solve this you'd need to make an exclusion for better-escape to not set the mappings when `filetype=fzf`.

Alternatively, you can probably use the `on_create` event to "unmap" `jj|jk` yourself.

**Reply by OriLiMu** [reply]
> > Thank you so much. I will try this one.
> 
> Let me know what you think after you try it out for a while, if this is an acceptable workaround I might add a new profile with it maybe we call it `telescope+normal`?



how to use `jk` to trigger the normal and insert mode? 
Like the telescope way to do it. And how to make it start as the insert mode.

**Reply by ibhagwan** [reply]
This has been asked many times, search the issues.
